integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonIngestCommand.scala

```diff
@@ -0,0 +1,368 @@
+      // We use a snapshot file to indicate whether there was failure in previous
+      // ingest operation. A Snapshot file will be created when an ingest operation
+      // starts and will be deleted only after the whole ingest operation is finished,
+      // which includes two actions:
+      //   1) action1: changing segment status to SUCCESS and
+      //   2) action2: deleting all involved stage files.
+      //
+      // If one of these two actions is failed, the snapshot file will be exist, so
+      // that recovery is needed.
+      //
+      // To do the recovery, do following steps:
+      //   1) Check if corresponding segment in table status is SUCCESS,
+      //      means deleting stage files had failed. So need to read the stage
+      //      file list from the snapshot file and delete them again.
+      //   2) Check if corresponding segment in table status is INSERT_IN_PROGRESS,
+      //      means data loading had failed. So need to read the stage file list
+      //      from the snapshot file and load again.
+      recoverIfRequired(snapshotFilePath, table, hadoopConf)
+
+      // 2. Start ingesting, steps:
+      //   1) read all existing stage files
+      //   2) read all stage files to collect input files for data loading
+      //   3) add a new segment entry in table status as INSERT_IN_PROGRESS,
+      //   4) write all existing stage file names into a new snapshot file
+      //   5) do actual loading
+      //   6) write segment file and update segment state to SUCCESS in table status
+      //   7) delete stage files used for loading
+      //   8) delete the snapshot file
+
+      // 1) read all existing stage files
+      val stageFiles = listStageFiles(stagePath, hadoopConf)
+      if (stageFiles.isEmpty) {
+        // no stage files, so do nothing
+        LOGGER.warn("files not found under stage metadata folder")
+        return Seq.empty
+      }
+
+      // 2) read all stage files to collect input files for data loading
+      // create a thread pool to read them
+      val numThreads = Math.min(Math.max(stageFiles.length, 1), 10)
+      val executorService = Executors.newFixedThreadPool(numThreads)
+      val stageInputs = collectStageInputs(executorService, stageFiles)
+
+      // 3) add new segment with INSERT_IN_PROGRESS into table status
+      val loadModel = DataLoadProcessBuilderOnSpark.createLoadModelForGlobalSort(spark, table)
+      CarbonLoaderUtil.recordNewLoadMetadata(loadModel)
+
+      // 4) write all existing stage file names and segmentId into a new snapshot file
+      // The content of snapshot file is: first line is segmentId, followed by each line is
+      // one stage file name
+      val content =
+        (Seq(loadModel.getSegmentId) ++ stageFiles.map(_._1.getAbsolutePath)).mkString("\n")
+      FileFactory.writeFile(content, snapshotFilePath)
+
+      // 5) perform data loading
+      startLoading(spark, table, loadModel, stageInputs)
```

Comment:
If Load Fails, still new load will be in `Insert_in_progress`. Needs to be cleaned/Marked for delete in finally block

Reply:
added handling in catch block